nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java

```diff
@@ -0,0 +1,169 @@
+          Nullness.hasNullableAnnotation(annotationMirrors.stream(), config);
+      if (hasNullableAnnotation) {
+        nullableTypeArguments.add(index);
+      }
+    }
+
+    Type baseType = type.tsym.type;
+    com.sun.tools.javac.util.List<Type> baseTypeArguments = baseType.getTypeArguments();
+    checkNullableTypeArgsAgainstUpperBounds(
+        state, tree, analysis, config, nullableTypeArguments, baseTypeArguments);
+    // Generics check for nested type parameters
+    for (Type typeArgument : typeArguments) {
+      if (typeArgument.getTypeArguments().length() > 0) {
+        checkInstantiatedType(typeArgument, state, tree, analysis, config);
+      }
+    }
+  }
+
+  /**
+   * Checks if the type arguments with a {@code @Nullable} annotation in an instantiated type have a
+   * {@code @Nullable} upper bound in the declaration, and reports an error otherwise.
+   *
+   * @param state the visitor state
+   * @param tree TODO not sure we need this
+   * @param analysis the analysis object
+   * @param config the analysis config
+   * @param nullableTypeArguments indices of {@code Nullable} type arguments
+   * @param baseTypeArguments list of type variables (with bounds) in the declared type
+   */
+  private static void checkNullableTypeArgsAgainstUpperBounds(
+      VisitorState state,
+      Tree tree,
+      NullAway analysis,
+      Config config,
+      HashSet<Integer> nullableTypeArguments,
+      com.sun.tools.javac.util.List<Type> baseTypeArguments) {
+    int index = 0;
+    for (Type baseTypeArg : baseTypeArguments) {
+
+      // if type argument at current index has @Nullable annotation base type argument at that
+      // index
+      // should also have a @Nullable annotation on its upper bound.
+      if (nullableTypeArguments.contains(index)) {
+
+        Type upperBound = baseTypeArg.getUpperBound();
+        com.sun.tools.javac.util.List<Attribute.TypeCompound> annotationMirrors =
+            upperBound.getAnnotationMirrors();
+        boolean hasNullableAnnotation =
+            Nullness.hasNullableAnnotation(annotationMirrors.stream(), config);
+        // if base type argument does not have @Nullable annotation then the instantiation is
+        // invalid
+        if (!hasNullableAnnotation) {
+          invalidInstantiationError(tree, state, analysis);
+        }
+      }
+      index++;
+    }
+  }
+
+  /**
+   * Checks that for an instantiated generic type, {@code @Nullable} types are only used for type
+   * variables that have a {@code @Nullable} upper bound. Similar to {@link
+   * #checkInstantiatedType(Type, VisitorState, Tree, NullAway, Config)} but specialized for when
+   * the instantiated type is represented as a {@link ParameterizedTypeTree}.
+   *
+   * @param tree the tree representing the instantiated type
+   * @param state visitor state
+   * @param analysis the analysis object
+   * @param config the analysis config
+   */
+  public static void checkInstantiationForParameterizedTypedTree(
```

Comment:
So, after further digging, I think we can do another simplification here.  The reason we wouldn't just use the `checkInstantiatedType` method everywhere is that sometimes, javac does not include type annotations when you get the `type` field of a `ParameterizedTypeTree`.  However, I think that wherever we currently call `checkInstantiatedType` from `NullAway`, there is also a `ParameterizedTypeTree` available; so we could just call `checkInstantiationForParameterizedTypedTree` instead.  I think that simplification is worthwhile for two reasons.  First, there are fewer code paths for the logic.  Second, and more importantly, I would like our error locations to point exactly to the place where the `@Nullable` annotation appears (see my other comment).  So let's aim for this change.